ddtrace/gitmetadata.py

```diff
@@ -0,0 +1,88 @@
+import os
+import inspect
+import subprocess
+import pkg_resources
+
+
+def __query_git(args):
+    try:
+        p = subprocess.Popen(["git"] + args, stdout=subprocess.PIPE)
+    except EnvironmentError:
+        print("Couldn't run git")
+        return
+    ver = p.communicate()[0]
+    return ver.strip().decode("utf-8")
+
+
+def get_commit_sha():
+    return __query_git(["rev-parse", "HEAD"])
+
+
+def get_repository_url():
+    return __query_git(["config", "--get", "remote.origin.url"])
+
+
+def add_project_urls(setup):
+    source_code_link = get_repository_url() + "#" + get_commit_sha()
+
+    def patch(**attrs):
+        if "project_urls" not in attrs:
+            attrs["project_urls"] = {}
+        attrs["project_urls"]["source_code_link"] = source_code_link
+        return setup(**attrs)
+
+    return patch
+
+
+def __find_package():
+    frm = inspect.stack()[-1]
+    module = inspect.getmodule(frm[0])
+    if module is None:
+        return None
+    path = module.__file__
+    package = ""
+    while len(path) > 1:
+        path, end = os.path.split(path)
+        if end in ("site-packages", "dist-packages"):
+            break
+        package = end
+    return package.split("-")[0]
+
+
+def get_tags(package=None):
+    if package is None:
+        package = __find_package()
+    if package is None:
+        return {}
+    try:
+        pkg = pkg_resources.require(package)[0]
+        metadata_lines = []
+        if pkg.has_metadata("METADATA"):
+            metadata_lines = pkg.get_metadata_lines("METADATA")
+        if pkg.has_metadata("PKG-INFO"):
+            metadata_lines = pkg.get_metadata_lines("PKG-INFO")
+
+        source_code_link = ""
+        for metadata in metadata_lines:
+            name_val = metadata.split(": ")
+            if name_val[0] == "Project-URL":
+                capt_val = name_val[1].split(", ")
+                if capt_val[0] == "source_code_link":
+                    source_code_link = capt_val[1].strip()
+                    break
```

Comment:
I _think_ there is a better way to do this, I cannot remember what it is, but leaving a comment as a reminder to research a better way.

`importlib` might parse this already, or else I think it is a valid MIME message, so can be parsed with the email module to get something easier to work with. (the performance of what you have is likely better than parsing)